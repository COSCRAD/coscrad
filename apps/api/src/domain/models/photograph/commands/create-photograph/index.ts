export * from './create-photograph.command';
export * from './create-photograph.command-handler';
export * from './photograph-created.event';
