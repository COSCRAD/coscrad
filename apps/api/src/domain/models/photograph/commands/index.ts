export * from './create-photograph';
export * from './photograph.commands.module';
