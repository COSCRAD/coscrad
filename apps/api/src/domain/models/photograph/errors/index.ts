export * from './invalid-mime-type-for-photograph.error';
