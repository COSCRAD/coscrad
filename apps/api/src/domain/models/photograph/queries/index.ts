export * from './photograph-query-repository.interface';
