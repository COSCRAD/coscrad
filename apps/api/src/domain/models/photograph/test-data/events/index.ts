export * from './get-photograph-test-event-builders';
