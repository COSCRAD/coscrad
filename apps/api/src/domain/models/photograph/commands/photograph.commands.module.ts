import { CommandModule } from '@coscrad/commands';
import { Module } from '@nestjs/common';
import { CreatePhotograph, CreatePhotographCommandHandler } from '.';
import { IdGenerationModule } from '../../../../lib/id-generation/id-generation.module';
import { PersistenceModule } from '../../../../persistence/persistence.module';

@Module({
    imports: [PersistenceModule, CommandModule, IdGenerationModule],
    providers: [
        CreatePhotographCommandHandler,
        ...[CreatePhotograph].map((Ctor) => ({
            provide: Ctor,
            useValue: Ctor,
        })),
    ],
    exports: [CreatePhotograph],
})
export class PhotographCommandsModule {}
