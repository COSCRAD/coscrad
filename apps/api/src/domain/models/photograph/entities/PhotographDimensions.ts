import { InternalError } from 'apps/api/src/lib/errors/InternalError';
import { PartialDTO } from 'apps/api/src/types/partial-dto';
import BaseDomainModel from '../../BaseDomainModel';

// This is a value-object
export default class PhotographDimensions extends BaseDomainModel {
    // Do we need the unit in the prop name? Do we need other units?
    readonly widthPX: number;

    readonly heightPX: number;

    constructor({ widthPX, heightPX }: PartialDTO<PhotographDimensions>) {
        super();

        this.widthPX = widthPX;

        this.heightPX = heightPX;
    }

    rescale(scaleFactor: number): PhotographDimensions {
        if (scaleFactor <= 0) {
            throw new InternalError(`Cannot scale photograph by invalid factor: ${scaleFactor}`);
        }

        return this.clone<PhotographDimensions>({
            widthPX: scaleFactor * this.widthPX,
            heightPX: scaleFactor * this.heightPX,
        });
    }
}
