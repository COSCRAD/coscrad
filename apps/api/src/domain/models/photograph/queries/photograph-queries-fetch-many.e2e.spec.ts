import {
    AggregateType,
    CoscradUserRole,
    HttpStatusCode,
    LanguageCode,
    MultilingualTextItemRole,
} from '@coscrad/api-interfaces';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import httpStatusCodes from '../../../../app/constants/httpStatusCodes';
import setUpIntegrationTest from '../../../../app/controllers/__tests__/setUpIntegrationTest';
import { clonePlainObjectWithOverrides } from '../../../../lib/utilities/clonePlainObjectWithOverrides';
import { ArangoCollectionId } from '../../../../persistence/database/collection-references/ArangoCollectionId';
import { ArangoDatabaseProvider } from '../../../../persistence/database/database.provider';
import generateDatabaseNameForTestSuite from '../../../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import TestRepositoryProvider from '../../../../persistence/repositories/__tests__/TestRepositoryProvider';
import buildTestDataInFlatFormat from '../../../../test-data/buildTestDataInFlatFormat';
import { TestEventStream } from '../../../../test-data/events';
import getValidAggregateInstanceForTest from '../../../__tests__/utilities/getValidAggregateInstanceForTest';
import { MultilingualText } from '../../../common/entities/multilingual-text';
import buildDummyUuid from '../../__tests__/utilities/buildDummyUuid';
import { AccessControlList } from '../../shared/access-control/access-control-list.entity';
import { CoscradUserWithGroups } from '../../user-management/user/entities/user/coscrad-user-with-groups';
import { CoscradUser } from '../../user-management/user/entities/user/coscrad-user.entity';
import { FullName } from '../../user-management/user/entities/user/full-name.entity';
import { PhotographCreated } from '../commands';
import {
    IPhotographQueryRepository,
    PHOTOGRAPH_QUERY_REPOSITORY_TOKEN,
} from './photograph-query-repository.interface';
import { PhotographViewModel } from './photograph.view-model';

const indexEndpoint = `/resources/photographs`;

// We require an existing user for ACL tests
const dummyQueryUserId = buildDummyUuid(4);

const { user: users } = buildTestDataInFlatFormat();

// TODO support user groups
const dummyUser = (users as CoscradUser[])[0].clone({
    authProviderUserId: `auth0|${dummyQueryUserId}`,
    id: dummyQueryUserId,
    roles: [CoscradUserRole.viewer],
});

// Set up test data (use event sourcing to set up state)
const photographTitle = `Photograph Name (in the language)`;

const originalLanguage = LanguageCode.Haida;

const photographTitleTranslation = `Photograph Name (in English)`;

const translationLanguage = LanguageCode.English;

const photographId = buildDummyUuid(101);

const photographCompositeIdentifier = {
    type: AggregateType.photograph,
    id: photographId,
};

const photographIdUnpublishedNoUserAccessId = buildDummyUuid(102);

const photographIdUnpublishedWithUserAccessId = buildDummyUuid(103);

const dummyContributorFirstName = 'Dumb';

const dummyContributorLastName = 'McContributor';

const dummyContributor = getValidAggregateInstanceForTest(AggregateType.contributor).clone({
    fullName: new FullName({
        firstName: dummyContributorFirstName,
        lastName: dummyContributorLastName,
    }),
});

const photographCreated = new TestEventStream().andThen<PhotographCreated>({
    type: 'PHOTOGRAPH_CREATED',
    payload: {
        title: photographTitle,
        languageCodeForTitle: originalLanguage,
    },
    meta: {
        // we also want to check that the contributions come through
        contributorIds: [dummyContributor.id],
    },
});

const dummyPhotographView = clonePlainObjectWithOverrides(
    PhotographViewModel.fromPhotographCreated(
        photographCreated.as(photographCompositeIdentifier)[0] as PhotographCreated
    ),
    {
        contributions: [
            {
                id: dummyContributor.id,
                fullName: dummyContributor.fullName.toString(),
            },
        ],
    }
);

const publicPhotographView = clonePlainObjectWithOverrides(dummyPhotographView, {
    isPublished: true,
    name: new MultilingualText(dummyPhotographView.name).translate({
        text: photographTitleTranslation,
        languageCode: translationLanguage,
        role: MultilingualTextItemRole.freeTranslation,
        // we are insisting by casting that the call to `translate` won't fail above
    }) as MultilingualText,
});

const privatePhotographThatUserCanAccess = clonePlainObjectWithOverrides(publicPhotographView, {
    id: photographIdUnpublishedWithUserAccessId,
    isPublished: false,
    accessControlList: new AccessControlList().allowUser(dummyQueryUserId),
});

const privatePhotographUserCannotAccess = clonePlainObjectWithOverrides(publicPhotographView, {
    id: photographIdUnpublishedNoUserAccessId,
    isPublished: false,
    // no special access
    accessControlList: new AccessControlList(),
});

describe(`when querying for a photograph: fetch many`, () => {
    const testDatabaseName = generateDatabaseNameForTestSuite();

    let app: INestApplication;

    let testRepositoryProvider: TestRepositoryProvider;

    let databaseProvider: ArangoDatabaseProvider;

    let photographQueryRepository: IPhotographQueryRepository;

    const setItUp = async (testUserWithGroups?: CoscradUserWithGroups) => {
        ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
            {
                ARANGO_DB_NAME: testDatabaseName,
            },
            {
                testUserWithGroups,
            }
        ));

        photographQueryRepository = app.get(PHOTOGRAPH_QUERY_REPOSITORY_TOKEN);
    };

    afterAll(async () => {
        await app.close();

        databaseProvider.close();
    });

    describe(`when the user is unauthenticated`, () => {
        beforeAll(async () => {
            ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                {
                    ARANGO_DB_NAME: testDatabaseName,
                }
                // no authenticated user
            ));

            await app.init();

            photographQueryRepository = app.get(PHOTOGRAPH_QUERY_REPOSITORY_TOKEN);
        });

        describe(`when there is a published photograph in the index view`, () => {
            beforeEach(async () => {
                await databaseProvider
                    .getDatabaseForCollection(ArangoCollectionId.contributors)
                    .clear();

                await databaseProvider.getDatabaseForCollection('photograph__VIEWS').clear();

                await testRepositoryProvider.getContributorRepository().create(dummyContributor);

                await photographQueryRepository.createMany([
                    publicPhotographView,
                    privatePhotographThatUserCanAccess,
                    privatePhotographUserCannotAccess,
                ]);
            });

            it(`should only return published photographs`, async () => {
                const res = await request(app.getHttpServer()).get(indexEndpoint);

                expect(res.status).toBe(httpStatusCodes.ok);

                const {
                    body: { entities },
                } = res;

                // TODO check that the actions are correct !!!!!
                expect(entities).toHaveLength(1);

                // Only one published result should come through from eventHistoryForMany
                const result = entities[0] as PhotographViewModel;

                expect(result.id).toBe(photographId);

                const _knownContributors = await testRepositoryProvider
                    .getContributorRepository()
                    .fetchMany();

                expect(
                    result.contributions.some(({ fullName, id: foundContributorId }) => {
                        return (
                            // TODO we will want a test helper for this comparison in case we decide to change format
                            fullName ===
                                `${dummyContributorFirstName} ${dummyContributorLastName}` &&
                            foundContributorId === dummyContributor.id
                        );
                    })
                ).toBe(true);
            });
        });

        describe(`when there are no published photographs`, () => {
            beforeEach(async () => {
                await databaseProvider
                    .getDatabaseForCollection(ArangoCollectionId.contributors)
                    .clear();

                await databaseProvider.getDatabaseForCollection('photograph__VIEWS').clear();

                await testRepositoryProvider.getContributorRepository().create(dummyContributor);

                /**
                 * Note that there is no ordinary system user authenticated for the request
                 * in this scenario. This is a public request.
                 */
                await photographQueryRepository.createMany([
                    privatePhotographUserCannotAccess,
                    privatePhotographThatUserCanAccess,
                ]);
            });

            it(`should return no photographs`, async () => {
                const res = await request(app.getHttpServer()).get(indexEndpoint);

                expect(res.status).toBe(httpStatusCodes.ok);

                expect(res.body.entities).toHaveLength(0);
            });
        });
    });

    describe(`when the user is authenticated as a non-admin user`, () => {
        beforeAll(async () => {
            await setItUp(
                new CoscradUserWithGroups(
                    dummyUser.clone({
                        roles: [CoscradUserRole.viewer],
                    }),
                    []
                )
            );
        });

        describe(`when there are some resources the user should not access`, () => {
            beforeEach(async () => {
                await databaseProvider.getDatabaseForCollection('photograph__VIEWS').clear();

                await databaseProvider
                    .getDatabaseForCollection(ArangoCollectionId.contributors)
                    .clear();

                await photographQueryRepository.createMany([
                    publicPhotographView,
                    privatePhotographThatUserCanAccess,
                    // This one should not be visible to an ordinary user
                    privatePhotographUserCannotAccess,
                ]);

                await testRepositoryProvider.getContributorRepository().create(dummyContributor);
            });

            it(`should not return private results`, async () => {
                const res = await request(app.getHttpServer()).get(indexEndpoint);

                expect(res.status).toBe(httpStatusCodes.ok);

                /**
                 * + private, but user in ACL
                 * - private, user not in ACL
                 * + published
                 */
                expect(res.body.entities).toHaveLength(2);
            });
        });
    });

    describe('when the user is a project admin', () => {
        const userRole = CoscradUserRole.projectAdmin;

        beforeAll(async () => {
            await setItUp(
                new CoscradUserWithGroups(
                    dummyUser.clone({
                        roles: [userRole],
                    }),
                    []
                )
            );
        });

        beforeEach(async () => {
            await databaseProvider.getDatabaseForCollection('photograph__VIEWS').clear();

            await databaseProvider
                .getDatabaseForCollection(ArangoCollectionId.contributors)
                .clear();

            await photographQueryRepository.createMany([
                publicPhotographView,
                privatePhotographThatUserCanAccess,
                privatePhotographUserCannotAccess,
            ]);

            await testRepositoryProvider.getContributorRepository().create(dummyContributor);
        });

        it(`should allow the user to access private resources`, async () => {
            const res = await request(app.getHttpServer()).get(indexEndpoint);

            const numberOfPrivatePhotographs = 2;

            const numberOfPublicPhotographs = 1;

            expect(res.status).toBe(HttpStatusCode.ok);

            expect(res.body.entities).toHaveLength(
                numberOfPrivatePhotographs + numberOfPublicPhotographs
            );

            /**
             * This is to catch breaking changes in the API contract with
             * the client. See the above comment for the corresponding detail
             * endpoint test case.
             */
            expect(res.body).toMatchSnapshot();
        });
    });

    describe('when the user is a COSCRAD admin', () => {
        const userRole = CoscradUserRole.superAdmin;

        beforeAll(async () => {
            await setItUp(
                new CoscradUserWithGroups(
                    dummyUser.clone({
                        roles: [userRole],
                    }),
                    []
                )
            );
        });

        beforeEach(async () => {
            await databaseProvider.getDatabaseForCollection('photograph__VIEWS').clear();

            await databaseProvider
                .getDatabaseForCollection(ArangoCollectionId.contributors)
                .clear();

            await photographQueryRepository.createMany([
                publicPhotographView,
                privatePhotographThatUserCanAccess,
                privatePhotographUserCannotAccess,
            ]);

            await testRepositoryProvider.getContributorRepository().create(dummyContributor);
        });

        it(`should allow the user to access private resources`, async () => {
            const res = await request(app.getHttpServer()).get(indexEndpoint);

            const numberOfPrivatePhotographs = 2;

            const numberOfPublicPhotographs = 1;

            expect(res.status).toBe(HttpStatusCode.ok);

            expect(res.body.entities).toHaveLength(
                numberOfPrivatePhotographs + numberOfPublicPhotographs
            );
        });
    });
});
