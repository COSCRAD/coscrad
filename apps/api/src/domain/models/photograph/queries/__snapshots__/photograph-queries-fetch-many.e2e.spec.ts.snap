// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when querying for a photograph: fetch many when the user is a project admin should allow the user to access private resources 1`] = `
{
  "entities": [
    {
      "accessControlList": {
        "allowedGroupIds": [],
        "allowedUserIds": [],
      },
      "actions": [
        {
          "description": "Make a resource visible to the public",
          "form": {
            "fields": [],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100101",
                "type": "photograph",
              },
            },
          },
          "label": "Publish Resource",
          "type": "PUBLISH_RESOURCE",
        },
        {
          "description": "creates a note about this particular resource",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource about which we are making a note",
                "label": "CompositeIdentifier",
                "name": "resourceCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the note",
                "label": "text",
                "name": "text",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are writing the note",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100101",
                "type": "photograph",
              },
            },
          },
          "label": "Create Note",
          "type": "CREATE_NOTE_ABOUT_RESOURCE",
        },
        {
          "description": "connects two resources with a note and context",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource to which you are making a connection",
                "label": "to member composite identifier",
                "name": "toMemberCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource from which you are making a connection",
                "label": "from member composite identifier",
                "name": "fromMemberCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the note",
                "label": "text",
                "name": "text",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are writing the note",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100101",
                "type": "photograph",
              },
            },
          },
          "label": "Connect Resources with Note",
          "type": "CONNECT_RESOURCES_WITH_NOTE",
        },
      ],
      "contributions": [
        {
          "fullName": "Dumb McContributor",
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110901",
        },
      ],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100101",
      "isPublished": true,
      "name": {
        "items": [
          {
            "languageCode": "hai",
            "role": "original",
            "text": "Photograph Name (in the language)",
          },
          {
            "languageCode": "en",
            "role": "free translation",
            "text": "Photograph Name (in English)",
          },
        ],
      },
    },
    {
      "accessControlList": {
        "allowedGroupIds": [],
        "allowedUserIds": [
          "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100004",
        ],
      },
      "actions": [
        {
          "description": "Make a resource visible to the public",
          "form": {
            "fields": [],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
                "type": "photograph",
              },
            },
          },
          "label": "Publish Resource",
          "type": "PUBLISH_RESOURCE",
        },
        {
          "description": "creates a note about this particular resource",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource about which we are making a note",
                "label": "CompositeIdentifier",
                "name": "resourceCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the note",
                "label": "text",
                "name": "text",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are writing the note",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
                "type": "photograph",
              },
            },
          },
          "label": "Create Note",
          "type": "CREATE_NOTE_ABOUT_RESOURCE",
        },
        {
          "description": "connects two resources with a note and context",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource to which you are making a connection",
                "label": "to member composite identifier",
                "name": "toMemberCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource from which you are making a connection",
                "label": "from member composite identifier",
                "name": "fromMemberCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the note",
                "label": "text",
                "name": "text",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are writing the note",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
                "type": "photograph",
              },
            },
          },
          "label": "Connect Resources with Note",
          "type": "CONNECT_RESOURCES_WITH_NOTE",
        },
      ],
      "contributions": [
        {
          "fullName": "Dumb McContributor",
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110901",
        },
      ],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
      "isPublished": false,
      "name": {
        "items": [
          {
            "languageCode": "hai",
            "role": "original",
            "text": "Photograph Name (in the language)",
          },
          {
            "languageCode": "en",
            "role": "free translation",
            "text": "Photograph Name (in English)",
          },
        ],
      },
    },
    {
      "accessControlList": {
        "allowedGroupIds": [],
        "allowedUserIds": [],
      },
      "actions": [
        {
          "description": "Make a resource visible to the public",
          "form": {
            "fields": [],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
                "type": "photograph",
              },
            },
          },
          "label": "Publish Resource",
          "type": "PUBLISH_RESOURCE",
        },
        {
          "description": "creates a note about this particular resource",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource about which we are making a note",
                "label": "CompositeIdentifier",
                "name": "resourceCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the note",
                "label": "text",
                "name": "text",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are writing the note",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
                "type": "photograph",
              },
            },
          },
          "label": "Create Note",
          "type": "CREATE_NOTE_ABOUT_RESOURCE",
        },
        {
          "description": "connects two resources with a note and context",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource to which you are making a connection",
                "label": "to member composite identifier",
                "name": "toMemberCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource from which you are making a connection",
                "label": "from member composite identifier",
                "name": "fromMemberCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the note",
                "label": "text",
                "name": "text",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are writing the note",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
                "type": "photograph",
              },
            },
          },
          "label": "Connect Resources with Note",
          "type": "CONNECT_RESOURCES_WITH_NOTE",
        },
      ],
      "contributions": [
        {
          "fullName": "Dumb McContributor",
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110901",
        },
      ],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
      "isPublished": false,
      "name": {
        "items": [
          {
            "languageCode": "hai",
            "role": "original",
            "text": "Photograph Name (in the language)",
          },
          {
            "languageCode": "en",
            "role": "free translation",
            "text": "Photograph Name (in English)",
          },
        ],
      },
    },
  ],
  "indexScopedActions": [],
}
`;
